### PR TITLE
Stop words

### LINKED ISSUE #111: Implement stop words
Implement stop words to make transplant iso with the current version of MeiliSearch

### LINKED ISSUE #150: Fix stop words
_Following https://github.com/meilisearch/transplant/pull/135#pullrequestreview-630336982_

I have an issue when using stop-words:
- add documents with `title: "the hobbit"`
- add `the` as stop-words
- search for "the hobbit"

-> I don't get any result, I should get my document with "the hobbit" if I'm not wrong.

### PR DESCRIPTION
**Pull request by irevoire**
Closes #111 and closes #150.

### CODE REVIEW

File: milli/src/search/mod.rs [hunk 1]
```diff
@@ -91,7 +91,13 @@ impl<'a> Search<'a> {
                 let mut builder = QueryTreeBuilder::new(self.rtxn, self.index);
                 builder.optional_words(self.optional_words);
                 builder.authorize_typos(self.authorize_typos);
-                let analyzer = Analyzer::<Vec<u8>>::new(AnalyzerConfig::default());
+                // We remove all the stop_words from the query
```
Comment (Kerollmops) [inline milli/src/search/mod.rs:94]:
```suggestion
                // We make sure that the analyzer is aware of the stop words
                // this ensures that the query builder is able to properly remove them.
```

File: milli/src/search/query_tree.rs [hunk 2]
```diff
@@ -229,7 +221,7 @@ impl<'a> QueryTreeBuilder<'a> {
     ///   forcing all query words to match documents without any typo
     ///   (the criterion `typo` will be ignored)
     pub fn build(&self, query: TokenStream) -> anyhow::Result<Option<Operation>> {
-        let primitive_query = create_primitive_query(query);
+        let primitive_query = create_primitive_query(query, self.index.stop_words(self.rtxn)?);
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:224]:
```suggestion
        let stop_words = self.index.stop_words(self.rtxn)?;
        let primitive_query = create_primitive_query(query, stop_words);
```

File: milli/src/search/query_tree.rs [hunk 3]
```diff
@@ -378,11 +366,12 @@ fn create_query_tree(
...
         authorize_typos: bool,
         query: &[PrimitiveQueryPart],
-    ) -> anyhow::Result<Operation> {
+    ) -> anyhow::Result<Operation>
+    {
         const MAX_NGRAM: usize = 3;
         let mut op_children = Vec::new();
 
-        for sub_query in query.linear_group_by(|a, b| !(a.is_phrase() || b.is_phrase())) {
+        for sub_query in query.linear_group_by(|a, b| !(a.is_phrase() || b.is_phrase()) ) {
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:374]:
```suggestion
        for sub_query in query.linear_group_by(|a, b| !(a.is_phrase() || b.is_phrase())) {
```

File: milli/src/search/query_tree.rs [hunk 4]
```diff
@@ -393,31 +382,23 @@ fn create_query_tree(
...
-                                .iter()
-                                .filter_map(|part| {
-                                    if let PrimitiveQueryPart::Word(word, _) = part {
-                                        Some(word.as_str())
-                                    } else {
-                                        None
-                                    }
-                                })
-                                .collect();
+                            let is_prefix = words.last().map(|part| part.is_prefix()).unwrap_or(false);
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:389]:
```suggestion
                            let is_prefix = words.last().map_or(false, |part| part.is_prefix());
```

File: milli/src/search/query_tree.rs [hunk 5]
```diff
@@ -393,31 +382,23 @@ fn create_query_tree(
...
+                            let words: Vec<_> = words.iter().filter_map(| part| {
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:390]:
```suggestion
                            let words: Vec<_> = words.iter().filter_map(|part| {
```

File: milli/src/search/query_tree.rs [hunk 6]
```diff
@@ -494,22 +475,24 @@ impl PrimitiveQueryPart {
...
+            TokenKind::Word | TokenKind::StopWord  => {
                 // 1. if the word is quoted we push it in a phrase-buffer waiting for the ending quote,
-                // 2. if the word is not the last token of the query we push it as a non-prefix word,
+                // 2. if the word is not the last token of the query and is not a stop_word we push it as a non-prefix word,
                 // 3. if the word is the last token of the query we push it as a prefix word.
                 if quoted {
                     phrase.push(token.word.to_string());
                 } else if peekable.peek().is_some() {
-                    primitive_query.push(PrimitiveQueryPart::Word(token.word.to_string(), false));
+                     if !stop_words.as_ref().map_or(false, |swords| swords.contains(token.word.to_string())) {
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:493]:
```suggestion
                     if !stop_words.as_ref().map_or(false, |swords| swords.contains(token.word)) {
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Could you please rebase on main, please?

File: milli/src/search/query_tree.rs [hunk 7]
```diff
@@ -155,10 +155,6 @@ impl fmt::Debug for Query {
 
 trait Context {
     fn word_docids(&self, word: &str) -> heed::Result<Option<RoaringBitmap>>;
-    fn stop_words(&self) -> anyhow::Result<Option<Set<&[u8]>>>;
-    fn is_stop_word(&self, word: &str) -> anyhow::Result<bool> {
-        Ok(self.stop_words()?.map_or(false, |s| s.contains(word)))
-    }
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:161]: I don't remember why these function have been removed from the context, and why we now use the index directly?

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: milli/src/search/query_tree.rs (hunk 7 above)
Comment (irevoire) [reply milli/src/search/query_tree.rs:161]: it was removed by the revert of the first commit.
It was added so we could access the stop_words in the `resolve_primitive_part`, but we don't need it now.

**Review by Kerollmops — APPROVED** [review]
Seems good to me, merging 🎉

**Pull request merged**